xsnap/sources/xsnap.c

```diff
@@ -484,7 +491,52 @@ void xs_print(xsMachine* the)
 	for (i = 0; i < c; i++) {
 		if (i)
 			fprintf(stdout, " ");
-		fprintf(stdout, "%s", xsToString(xsArg(i)));
+		xsArg(i) = xsCallFunction1(xsVar(0), xsUndefined, xsArg(i));
+		p = string = xsToString(xsArg(i));
+	#if mxCESU8
+		for (;;) {
+			xsIntegerValue character;
+			q = fxUTF8Decode(p, &character);
+		again:
+			if (character == C_EOF)
+				break;
+			if (character == 0) {
+				if (p > string) {
+					char c = *p;
+					*p = 0;
+					fprintf(stdout, "%s", string);
+					*p = c;
+				}
+				string = q;
+			}
+			else if ((0x0000D800 <= character) && (character <= 0x0000DBFF)) {
+				xsStringValue r = q;
+				xsIntegerValue surrogate;
+				q = fxUTF8Decode(r, &surrogate);
+				if ((0x0000DC00 <= surrogate) && (surrogate <= 0x0000DFFF)) {
+					char buffer[5];
+					character = (xsIntegerValue)(0x00010000 + ((character & 0x03FF) << 10) + (surrogate & 0x03FF));
+					if (p > string) {
+						char c = *p;
+						*p = 0;
+						fprintf(stdout, "%s", string);
+						*p = c;
+					}
+					p = fxUTF8Encode(buffer, character);
+					*p = 0;
+					fprintf(stdout, "%s", buffer);
+					string = q;
+				}
+				else {
+					p = r;
+					character = surrogate;
```

Comment:
Is there an opportunity to factor out a utility here? I imagine this will be brittle under maintenance.

Reply:
Yes. We'd prefer to wait a bit, to better understand what is really needed.